File: heron/common/src/cpp/config/cluster-config-reader.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "yaml-cpp/yaml.h"
#include "proto/messages.h"

#include "basics/basics.h"
#include "errors/errors.h"
#include "threads/threads.h"
#include "network/network.h"

#include "config/yaml-file-reader.h"
#include "config/cluster-config-vars.h"
#include "config/cluster-config-reader.h"

namespace heron {
namespace config {

ClusterConfigReader::ClusterConfigReader(std::shared_ptr<EventLoop> eventLoop,
        const sp_string& _defaults_file): YamlFileReader(eventLoop, _defaults_file) {
  LoadConfig();
}

ClusterConfigReader::~ClusterConfigReader() {}

void ClusterConfigReader::FillClusterConfig(proto::api::Topology* _topology) {
  // Fill in the cluster config
  for (YAML::const_iterator iter = config_.begin(); iter != config_.end(); ++iter) {
    proto::api::Config::KeyValue* kv = _topology->mutable_topology_config()->add_kvs();
    kv->set_key(iter->first.as<sp_string>());
    kv->set_value(iter->second.as<sp_string>());
  }
}

void ClusterConfigReader::OnConfigFileLoad() {
  AddIfMissing(ClusterConfigVars::CLUSTER_METRICS_INTERVAL, "60");
}
}  // namespace config
}  // namespace heron

File: heron/common/src/java/com/twitter/heron/common/config/ClusterConfigReader.java
// Copyright 2016 Twitter. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.twitter.heron.common.config;

import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

/**
 * Loads config file in Yaml properties file format based
 * on cluster and config path
 */
public final class ClusterConfigReader extends ConfigReader {
  private static final Logger LOG = Logger.getLogger(ClusterConfigReader.class.getName());

  public static Map<String, Object> load(String cluster, String configPath, String fileName) {
    Map<String, Object> config = new HashMap<>();

    // Read the defaults file, first
    String file1 = Paths.get(configPath, fileName).toString();
    LOG.info("Cluster " + cluster + " config file " + file1 + "\n");
    Map<String, Object> props1 = loadFile(file1);

    if (props1 == null) {
      LOG.info("props1 is null \n");
    }

    if (props1.isEmpty()) {
      LOG.info("Config file " + file1 + " is empty");
    }

    // Read the cluster specific file and override
    String file2 = Paths.get(configPath, cluster, fileName).toString();
    Map<String, Object> props2 = loadFile(file2);
    if (props2.isEmpty()) {
      LOG.info("Config file " + file2 + " is empty");
    }

    // If both files have errors, return false
    if (props1.isEmpty() && props2.isEmpty()) {
      LOG.info("Files " + file1 + " and " + file2 + " are not present");
      return config;
    }

    // Save the properties and return successfully
    config.putAll(props1);
    config.putAll(props2);
    return config;
  }
}
